project_2/include/packet.hpp

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <cstdint>
+#include <ostream>
+
+/**
+ * @brief Network packet abstraction
+ * 
+ */
+class PacketHeader {
+public:
+    std::uint64_t source_id;
+    std::uint64_t destination_id;
+    // TODO: use it as real sending time (passing first link maybe) in checks that
+    std::uint64_t sending_time;
+    std::uint32_t packet_index;
+    std::uint32_t size;
+
+    // 0 bit - is credit packet
+    // 2 bit - initializing connection or not
```

Comment:
I see you introduced a flag on bit 2 for handshaking. Do we actually need handshaking for the credit-based congestion contol?

Reply:
Yes, when sender wants to send some data it has to setup connection with receiver with such packet
From the paper:
![image](https://github.com/user-attachments/assets/a238eb58-00d9-4e1a-8b78-0611b6b60a44)
Starting and stopping credit flow (end-host): Finally, ExpressPass requires a signaling mechanism to start the credit flow at the receiver. For TCP, we piggyback credit request to either SYN and/or SYN+ACK packet depending on data availability. This allows data transmission to start immediately after the 3-way handshake. For persistent connections or UDP, we send credit requests in a minimum sized packet, but this adds a delay of an RTT.

Reply:
Thanks for explaining. Yes, this is how it works in the real system. However, for the sake of simulation and modelling, can we assume that the connection is established and manage this within the Flow class? Sort of: 

Flow created => start sending credits; Flow deleted => stop sending credits. 

Although, this is not exact representation of the real system, this is simpler and might be sufficient for the purpose of congestion control simulations.

Reply:
I think yes, we can assume that and it would be enougth in our case